src/Symfony/Component/HttpFoundation/JsonResponse.php

```diff
@@ -90,16 +90,52 @@ public function setCallback($callback = null)
      * @param mixed $data
      *
      * @return JsonResponse
-     *
-     * @throws \InvalidArgumentException
      */
     public function setData($data = array())
     {
+        $this->data = $data;
```

Comment:
you are still breaking BC by storing the array/object/scalar instead of storing the JSON string.

And it might even break things in case your data is a mutable object, as your data could be mutated before you trigger another update